content/courses/advanced-ada/chapters/access_types.rst

```diff
@@ -294,14 +294,430 @@ Unchecked Access
+access value |mdash| which was previously pointing to an existing object
+|mdash| is set to :ada:`null`. Therefore, :ada:`I = null` after the call to
+:ada:`Free`, which is exactly what we see when running this example code.
+
+Note that it is OK to call :ada:`Free` multiple times for the same access
+object:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Unchecked_Deallocation
+
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I : Integer_Access;
+
+    begin
+       I := new Integer;
+
+       Free (I);
+       Free (I);
+       Free (I);
+    end Show_Unchecked_Deallocation;
+
+The multiple calls to :ada:`Free` for the same access object don't cause any
+issues. Because the access value is null after the first call to
+:ada:`Free (I)`, we're actually just passing :ada:`null` as an argument in the
+second and third calls to :ada:`Free`. However, any attempt to deallocate an
+access value of null is ignored in the :ada:`Free` procedure, so the second and
+third calls to :ada:`Free` don't have any effect.
+
+.. admonition:: In the Ada Reference Manual
+
+    - `4.8 Allocators <https://www.adaic.org/resources/add_content/standards/12rm/html/RM-4-8.html>`__
+    - `13.11.2 Unchecked Storage Deallocation <https://www.adaic.org/resources/add_content/standards/12rm/html/RM-13-11-2.html>`__
+
+
+Dangling References
+~~~~~~~~~~~~~~~~~~~
+
+An access value that points to a non-existent object is called a dangling
+reference. In this section, we discuss the issues of having dangling
+references.
+
+Let's reuse the last example and introduce :ada:`I_2`, which will point to the
+same object as :ada:`I`:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Access_Types.Unchecked_Deallocation
+    :class: ada-run-expect-failure
+
+    with Integer_Types; use Integer_Types;
+
+    procedure Show_Unchecked_Deallocation is
+
+       I, I_2 : Integer_Access;
+
+    begin
+       I := new Integer;
+
+       I_2 := I;
+
+       --  NOTE: I_2 points to the same
+       --        object as I.
+
+       --
+       --  Use I and I_2...
+       --
+       --  ... then deallocate memory...
+       --
 
-    - `Unchecked Storage Deallocation <http://www.ada-auth.org/standards/2xrm/html/RM-13-11-2.html>`_
+       Free (I);
+
+       --  NOTE: at this point, I_2 is a
+       --        dangling reference!
+
+       --  Further calls to Free (I)
+       --  are OK!
+
+       Free (I);
+       Free (I);
+
+       --  A call to Free (I_2) is
+       --  NOT OK:
+
+       Free (I_2);
+    end Show_Unchecked_Deallocation;
+
+As we've seen before, we can have multiple calls to :ada:`Free (I)`.
+However, the call to :ada:`Free (I_2)` is bad because :ada:`I_2` is not null.
+In fact, it is a dangling reference |mdash| i.e. :ada:`I_2` points to an object
+that doesn't exist anymore. Also, the previous call to :ada:`Free (I)` will try
```

Comment:
'Also, the previous call to " should probably be "Also, the first call to " because the previous call is the third call for I, which has no effect.

Reply:
Good catch, thanks!